Final draft of Lucky Star blog post (#47)
dagster-io/dagster-website: pages/blog/dagster-0-11-0-lucky-star.mdx

excerpt: 'In 0.10.0, we introduce dynamic orchestration, a new backfill UI, and support for tracking asset lineage.'
date: '2021-04-01'
tags: ['dagster']
---

Please welcome to the stage our eleventh major release of Dagster, codenamed "Lucky Star."

<img
  align="right"
  src="/posts/dagster-0-11-0-lucky-star/lucky_star.png"
  alt="Dagster is coiffed to impress"
/>

This release marks a substantial step forwards in several areas.

### Release highlights
* *Dynamic orchestration*, which lets pipeline subsets execute once for each element of an input processed at runtime, is now ready for wider usage.
* *Backfills* of pipeline runs over partitioned datasets, to rebuild data assets when processes or underlying data change, are now supported with specialized UI. 
* *Asset lineage* to track the interdependencies of data assets produced by pipeline runs now has core API support.

We've also launched a new docs site, improved the asset catalog, improved interoperability between the Dagster type system and Python type hints, made many improvements to Dagit and to the robustness of core system components, and now support MySQL as a backend for core storage.

# Dynamic orchestration

> Many data pipelines include steps that have to be repeated once for each element of some input. Very often, you don't know in advance how many inputs youre going to have to process. And once you're done, you'll often want to collect the outputs of each process for downstream processing as an aggregate.

Reviewer: ```suggestion
Many data pipelines include steps that have to be repeated once for each element of some input. Very often, you don't know in advance how many inputs you're going to have to process. And once you're done, you'll often want to collect the outputs of each process for downstream processing as an aggregate.
```

Author: thx

For example, if you're processing files that you grab from an SFTP drop, you may want to run a complex subset of a pipeline once for each file, but the number of files may differ day-to-day. Once you've processed all the files, you might want to compute aggregate statistics or create a summary report.

> From the standpoint of the orchestration engine, we think of this as the simplest kind of _dynamic orchestration_ -- where the work to be orchestrated is determined not at pipeline definition time but at runtime, dependent on data that's observed as part of pipeline execution.

Reviewer 2: "From the standpoint of the orchestration engine, we think of this as the simplest kind of _dynamic orchestration_"

This is a mouthful. Not sure what we are trying to say here exactly. 

Seems like it would be just "Dagster thinks of this as  _dynamic orchestration_":

Author: ok

One of our [longest-standing Github issues](https://github.com/dagster-io/dagster/issues/462) has been for an abstraction to help model this common case. In Dagster 0.10.0 "The Edge of Glory", we introduced experimental support for dynamic orchestration with the `map` operation, allowing solids to yield special outputs using the [`DynamicOutput`](https://docs.dagster.io/_apidocs/dynamic#dagster.experimental.DynamicOutput) class, over which the execution of downstream solids could be mapped at runtime.

> In 0.11.0, we've expanded our experimental support for this pattern (and closed issue 462!) with the `collect` operation, so that downstream solids can aggregate and process all of the results of the solids executed dynamically using map.

Reviewer: ```suggestion
In 0.11.0, we've expanded our experimental support for this pattern (and closed issue 462!) with the `collect` operation so that downstream solids can aggregate and process all of the results of the solids executed dynamically using `map`.
```

Author: thx

```python
@solid(
    config_schema={"path": str},
    output_defs=[DynamicOutputDefinition(str)],
)
def files_in_directory(context):
    path = context.solid_config["path"]
    dirname, _, filenames = next(os.walk(path))
    for file in filenames:
        yield DynamicOutput(
            value=os.path.join(dirname, file),
            mapping_key=file_name_to_key(file),
        )
```

In the pipeline definition, we need to map over the outputs of this solid and then, later, collect the results of the mapped computation.

```python
@pipeline
def process_directory():
    files = files_in_directory()
    file_results = files.map(process_file)
    summarize_directory(file_results.collect())
```

> Since the `map` and `collect` operations are part of the pipeline definition, we can visualize them in Dagit. Both the mapped dynamic output and the collected input have special sigils and the mapped solid is embellished in the pipeline view to illustrate that it will be executed dynamically. 

Reviewer 2: I would add:

"We believe deeply in always rendering a pipeline in its logical form _before_ computation, even in the presence of dynamic orchestration. A user should know what is going to happen before a pipeline executes."

^----- 

This point is really unique to our view of the world, and I think we should emphasize it whenever we can.

Author: ok

<p align="center">
  <img
    src="/posts/dagster-0-11-0-lucky-star/dynamic_pipeline.png"
    alt="Dynamic pipeline as it appears in Dagit"
    width="700px"
  />
</p>

When we execute the pipeline, each copy of the mapped computations is identified in the execution timeline view by its associated mapping_key in square brackets.

<p align="center">
  <img
    src="/posts/dagster-0-11-0-lucky-star/dynamic_orchestration.gif"
    alt="Dynamic pipeline in Dagit's execution viewer"
  />
</p>

> Keep in mind that Dagster is not a specialized distributed computation framework like [Dask](https://dask.org/) or [Spark](https://spark.apache.org/). If you need to process very large datasets efficiently in parallel, we still recommend that you use Dagster to orchestrate solids that perform compute in a specialized framework. But for many coarse-grained cases, these abstractions make it easier to express logical parallelism and write pipelines that get work done.

Reviewer 2: This is oddly worded, as if straddling the line between recommending using Dagster and not and it's up coming off strangely.

"Dynamic orchestration is not meant to replace frameworks that perform massively parallel computation over large data sets. For that task use Spark, Dask, or similar. Dagster is at a different layer of the stack, and dynamic orchestration is for coarse-grained, logical parallelism in pipelines."

Author: ok

We'd love to learn about your use cases and to get your feedback on this feature as we move it from experimental to fully supported.

# Running backfills from Dagit

As data practitioners, we often find ourselves working with _partitioned_ data pipelines. For example, you might have a pipeline that consumes a table of events generated by a web application and produces aggregate statistics, like "average session length." It's quite likely that you'll want to partition this pipeline by time -- perhaps running it daily, over the past day's data -- so that you can track the aggregate statistics as they change over time. Or you may have a pipeline that consumes a table of user data, where you run the pipeline for each user cohort to track the performance of different user groups.

> But partitions can lead to operational complexities. When things can going smoothly, you can just incrementally update your data assets by creating new pipeline runs as new data becomes available. But what happens when you change the definition of a key metric and want to compare it to historical data that was computed using different pipeline logic? What if your upstream data is revised and you need to understand the consequences? This is where it becomes critical to kick off a _backfill_ -- a set of pipeline runs that revise previously calculated partitions.

Reviewer: ```suggestion
But using partitions can lead to operational complexities. When things can go smoothly, you can incrementally update your data assets by creating new pipeline runs as new data becomes available. But what happens when you change the definition of a key metric, and you need to compare it to historical data that was computed using different pipeline logic? What if your upstream data is revised and you need to understand the consequences? This is where it becomes critical to kick off a _backfill_ -- a set of pipeline runs that revise previously calculated partitions.
```

What is "things"?
things?

Author: cool this is just gibberish

We're continuing to invest in making this task easier than ever, and in 0.11.0, we're shipping a new dedicated backfill interface in Dagit to help you launch, monitor, and cancel in-flight backfills over any or all of your partitions.

## The Partitions Tab

In Dagster, partitioned pipelines are represented using the [`PartitionSet`](https://docs.dagster.io/concepts/partitions-schedules-sensors/partitions) API, which defines the partitions over which a given pipeline can be run (for example, a set of days).

> The Dagit partitions tab makes it easy to view all the historical runs for a particular partition set, organized by partition. You can navigate to this interface from the pipeline page, or by searching for a partition set in Dagit's universal search box.

Reviewer: ```suggestion
The Dagit partitions tab makes it easy to view all the historical runs for a particular partition set, organized by partition. You can navigate to this interface from the pipeline page or search for a partition set in Dagit's universal search box.
```

<p align="center">
  <img
    src="/posts/dagster-0-11-0-lucky-star/partitions_tab.png"
    alt="The Dagit partitions tab"
    width="900px"
  />
</p>

> Each row in this matrix corresponds to one solid in the partitioned pipeline, and each column corresponds to one of the partitions, so each box represents the execution of a solid over a particular partition. The shade of the box indicates the state of the _latest_ run for that solid over the given partition. Red indicates a failure, and white indicates that no run has executed.

Reviewer: ```suggestion
Each row in this matrix corresponds to one solid in the partitioned pipeline. Each column corresponds to one of the partitions, so each box represents a solid's execution over a particular partition. The box's shade indicates the state of the _latest_ run for that solid over the given partition. Red indicates a failure, and white indicates that no run has been executed.
```

> You can drill down into an individual partition to see the history of all runs executed for that partition:

Reviewer 2: Something akin to "Dagster keeps an immutable record of all runs, pipelines, and executions every seen by the system. This what enables us to list all runs that every touched a particular partition. Contrast this to Airflow, which destructively alters task history to retry parts of backfills."

I'd love to use this opportunity to really hammer home that this is a unique property. @sryza may be able to provide a more accurate wording for the airflow comparison

Author: yep took a crack but eager for @sryza's input

Reviewer 3: I like this part of Nick's suggestion: "Dagster keeps an immutable record of all runs, pipelines, and executions ever seen by the system. This enables us to list all runs that ever touched a particular partition."

If we're going to contrast with Airflow, IMO it's helpful to talk in terms of something more concrete than "a pleasant surprise".  Maybe something like, "Compared to when using systems like Airflow, which overwrite task history, this makes it possible to reconstruct what happened in the past when debugging tricky production issues".

<p align="center">
  <img
    src="/posts/dagster-0-11-0-lucky-star/partition_history.png"
    alt="The history of runs for a partition"
    width="900px"
  />
</p>

## Launching backfills

You can also launch backfills directly from the partitions tab. The backfill launch UI includes the ability to visually select partition to backfill:

<p align="center">
  <img
    src="/posts/dagster-0-11-0-lucky-star/launch_backfill.png"
    alt="Selecting the partitions over which to launch a backfill"
    width="900px"
  />
</p>

After launching a backfill, you can jump back to the partitions view to monitor its progress:

<p align="center">
  <img
    src="/posts/dagster-0-11-0-lucky-star/backfill_progress.png"
    alt="Observing the progress of a backfill"
    width="900px"
  />
</p>

## Execution model

Backfill execution should also be much more robust, with backfills now managed by Dagster's daemon process. This means you can launch large backfills over thousands of partitions without the risk of bringing Dagit down or interfering with other uses of the system.

> We know that backfills are often a huge pain for users of other data orchestration systems, so we're very interested in reducing the pain associated with this common chore. If you find yourself running a lot of backfills, the core team would greatly appreciate your feedback on usability.

Reviewer 2: We don't have to go into it here, but I would frame -- in a different article -- backfills as an unsolved "challenge problem" that the industry has of yet been unable to solve satisfactorily.

Author: yes, backfills would be a great article

You can associate outputs with assets in a couple of ways (see the [docs](https://docs.dagster.io/concepts/assets/asset-materializations#asset-lineage) for details), but the simplest is just to set the `asset_key` parameter on an [OutputDefinition](https://docs.dagster.io/_apidocs/solids#dagster.OutputDefinition).

```python
@solid(output_defs=[OutputDefinition(asset_key=AssetKey("my_db.users"))])
def scrape_users(_):
    users_df = some_api_call()
    persist_to_db(users_df)
    return users_df

@solid(output_defs=[OutputDefinition(asset_key=AssetKey("ml_models.user_prediction"))])
def create_model(_, users_df):
    my_ml_model = train_prediction_model(users_df)
    persist_to_model_store(my_ml_model)
    return my_ml_model

@pipeline
def my_user_model_pipeline():
    get_prediction_model(scrape_users())
```

In this toy example, our `scrape_users` solid pulls some data into a database table so that the downstream `create_model` solid can build an ML model. We'd like users of our data platform who are browsing through the asset catalog to know that the ML model depends on the database table.

> We've already told Dagster about the data dependencies between the inputs and outputs of our solids -- that's how we constructed the pipeline. When we attach asset keys to the outputs of our solids, we give Dagster the information it needs to infer the lineage graph that connects the associated assets.

Reviewer 2: These sentences are a mouthful.

"Dagster already knows about data dependencies between computations. With a minimal amount of metadata, Dagster can construct the lineage of the actual assets produced by those computations at runtime."

Author: ok

> <p align="center">

Reviewer 2: The arrows being the wrong way makes this diagram _very_ confusing. ml_model depends on db.user_table. The arrow should reflect that

Author: i'm happy to change this (will add a label) but it makes the diagram "noncommutative" which is always confusing; having one arrow (in the lineage graph) represent `depends_on` but another arrow (in the pipeline) represent the opposite is another source of confusion; we know from our interviews that even many of our engineering candidates find `depends_on` extremely counterintuitive

Author: how are we envisioning displaying the lineage graph in Dagit? I had imagined that the polarity of the edges would be the same as with the pipelines -- where root/upstream assets appeared at the "top" of the display and arrows flowed "down" to leaf/downstream assets. this seems to imply the opposite?

Author: ![Screen Shot 2021-04-02 at 8 24 39 AM](https://user-images.githubusercontent.com/427016/113429128-df8a0100-938c-11eb-9e2c-b7298c693107.png)

Author: replaced with this new diagram

Reviewer 2: i think with the text this is a good resolution for now

Reviewer 3: fwiw, when I google-searched for data lineage, most of the systems seemed to represent it in the direction of dataflow, vs. the direction of independence

some examples:
![image](https://user-images.githubusercontent.com/654855/113434088-34ca1080-9395-11eb-909d-ae6a6ca932f0.png)

![image](https://user-images.githubusercontent.com/654855/113434130-43182c80-9395-11eb-8c34-e1e066cde544.png)

![image](https://user-images.githubusercontent.com/654855/113434175-50351b80-9395-11eb-84e7-7568f9258b9a.png)

  <img
    src="/posts/dagster-0-11-0-lucky-star/lineage_graph.png"
    alt="Inferring the lineage graph from data dependencies"
    width="500px"
  />
</p>

It's early days for this feature, and we expect that the lineage interface in particular will evolve significantly. But as soon as you associate assets with outputs, you'll start to see information about asset lineage appear in the Asset Catalog in Dagit.

<p align="center">
  <img
    src="/posts/dagster-0-11-0-lucky-star/asset_lineage.png"
    alt="Asset lineage displayed in the Dagit asset catalog"
    width="600px"
  />
</p>

> We're very excited to see the use cases that the community serves with this feature, and to hear more about your speciifc needs: from data debugging (which tables do I need to query to find the source of my data issue?), to data discovery (which tables are used to train this model?), to impact analysis (if I change this file on s3, who do I need to notify?).

Reviewer: ```suggestion
We're very excited to see the use cases that the community serves with this feature, and to hear more about your specific needs: from data debugging (which tables do I need to query to find the source of my data issue?), to data discovery (which tables are used to train this model?), to impact analysis (if I change this file on s3, who do I need to notify?).
```

Check out the [docs](https://docs.dagster.io/concepts/assets/asset-materializations#asset-lineage) to learn more, and if you're excited about this product direction, please don't hesitate to reach out to the core team with feedback.

# New documentation site

> We're also very proud to present our new documentation site, redesigned from the underlying infrastructure to the information architecture of each page, which lives at [https://docs.dagster.io](https://docs.dagster.io/). This new site will host documentation for all Dagster versions from 0.11.0 onwards. (If you're on an older version of Dagster, you can still view pre-0.11.0 documentation at [https://legacy-docs.dagster.io](https://legacy-docs.dagster.io/).)

Reviewer: ```suggestion
We're also very proud to present our new documentation site, redesigned from the underlying infrastructure to each page's information architecture, which lives at [https://docs.dagster.io](https://docs.dagster.io/). This new site will host documentation for all Dagster versions from 0.11.0 onwards. (If you're on an older version of Dagster, you can still view pre-0.11.0 documentation at [https://legacy-docs.dagster.io](https://legacy-docs.dagster.io/).)
```

* [**Integrations**](https://docs.dagster.io/integrations) includes guides to the many integrations with popular data tools and platforms maintained as part of the core repository, including Pandas, dbt, Slack, Jupyter Notebooks, and more.

* The [**Guides**](https://docs.dagster.io/guides) are still early in development, but is intended to document best practices combining multiple Dagster concepts to accomplish common goals. If you have suggestions on what guides you would like to see, feel free to open an issue or send feedback on Slack.

* Finally, the [**API Reference**](https://docs.dagster.io/_apidocs)* contains comprehensive API documentation of all the decorators, functions, and classes exported by Dagster. If you want to explore the full API surface area, this is the place to search. 

## Searching the docs

> <p align="center">

Reviewer 2: This image is broken in the preview

Reviewer: @mgasner This needs to be `docs_search.gif`

Author: thx

  <img

>     src="/posts/dagster-0-11-0-lucky-star/docs_search.png"

Reviewer: ```suggestion
    src="/posts/dagster-0-11-0-lucky-star/docs_search.gif"
```
